packages-exp/auth-exp/src/core/user/proactive_refresh.ts

```diff
@@ -0,0 +1,99 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { User } from '../../model/user';
+import { AuthErrorCode } from '../errors';
+
+// Refresh the token five minutes before expiration
+export const _OFFSET_DURATION = 5 * 1000 * 60;
+
+export const _RETRY_BACKOFF_MIN = 30 * 1000;
+export const _RETRY_BACKOFF_MAX = 16 * 60 * 1000;
+
+export class ProactiveRefresh {
+  private isRunning = false;
+
+  // Node timers and browser timers return fundamentally different types.
```

Comment:
i have a similar issue in my PR, do you think it's worth extracting the setInterval logic into a common util?

Reply:
Eh that would mean adding more code that is compiled deadweight just to circumvent the type system.